src/pudl/transform/ferc1.py

```diff
@@ -42,6 +44,67 @@
 logger = pudl.logging_helpers.get_logger(__name__)
 
 
+@asset
+def clean_xbrl_metadata_json(
+    raw_xbrl_metadata_json: dict[str, dict[str, list[dict[str, Any]]]]
+) -> dict[str, dict[str, list[dict[str, Any]]]]:
+    """Generate cleaned json xbrl metadata.
+
+    For now, this only runs :func:`add_source_table_to_xbrl_metadata`.
+    """
+    return add_source_table_to_xbrl_metadata(raw_xbrl_metadata_json)
+
+
+def add_source_table_to_xbrl_metadata(
+    raw_xbrl_metadata_json: dict[str, dict[str, list[dict[str, Any]]]]
+) -> dict[str, dict[str, list[dict[str, Any]]]]:
+    """Add a ``source_table`` field into metadata calculation components.
+
+    When a particular component of a calculation does not originate from the table in
+    which the calculated field is being reported, label the source table.
+    """
+
+    def all_fields_in_table(table_meta) -> list:
+        """Compile a list of all of the fields reported in a table."""
+        return [
+            field["name"] for meta_list in table_meta.values() for field in meta_list
+        ]
+
+    def extract_tables_to_fields(xbrl_meta: dict) -> dict[str : list[str]]:
+        """Compile a dictionary of table names (keys) to list of fields."""
+        return {
+            table_name: all_fields_in_table(table_meta)
+            for table_name, table_meta in xbrl_meta.items()
+        }
+
+    def label_source_table(calc_component: dict, tables_to_fields: str) -> dict:
+        """Add a ``source_table`` element to the calculation component."""
+        calc_component["source_table"] = [
+            other_table_name
+            for other_table_name, fields in tables_to_fields.items()
+            if calc_component["name"] in fields
+        ]
+        # weirdly there are a number of nuclear xbrl_factoid calc components that seem
+        # to have no source_table.
+        if not calc_component["source_table"]:
+            logger.debug(f"Found no source table for {calc_component['name']}.")
+        return calc_component
+
+    tables_to_fields = extract_tables_to_fields(raw_xbrl_metadata_json)
+    # for each table loop through all of the calculations within each field
+    for table_name, table_meta in raw_xbrl_metadata_json.items():
+        for list_of_facts in table_meta.values():
+            for xbrl_fact in list_of_facts:
+                # all facts have ``calculations``, but they are empty lists when null
+                for calc_component in xbrl_fact["calculations"]:
+                    # does the calc component show up in the table? if not, add a label
+                    if calc_component["name"] not in tables_to_fields[table_name]:
+                        calc_component = label_source_table(
+                            calc_component, tables_to_fields
+                        )
```

Comment:
there has got to be a better way to do this.

Reply:
hard agree. i just don't think rn is the time to do this unless this becomes a source of friction.